src/game/client/ingameui.cpp

```diff
@@ -1751,3 +1765,1227 @@ void InGameUI::Recreate_Control_Bar()
+
+            if (alpha != 0) {
+                it++;
+                continue;
+            } else {
+                it = m_floatingTextList.erase(it);
+                data->Delete_Instance();
+            }
+        }
+    }
+}
+
+void InGameUI::Popup_Message(
+    Utf8String const &message, int width_percent, int height_percent, int width, bool pause, bool pause_music)
+{
+    Popup_Message(message, width_percent, height_percent, width, m_popupMessageColor, pause, pause_music);
+}
+
+void InGameUI::Popup_Message(
+    Utf8String const &message, int width_percent, int height_percent, int width, int color, bool pause, bool pause_music)
+{
+    if (m_popupMessageData != nullptr) {
+        Clear_Popup_Message_Data();
+    }
+
+    Update_Diplomacy_Briefing_Text(message, false);
+    Utf16String text = g_theGameText->Fetch(message);
+    m_popupMessageData = new PopupMessageData();
+    m_popupMessageData->m_message = text;
+
+    if (width_percent > 100) {
+        width_percent = 100;
+    }
+
+    if (width_percent < 0) {
+        width_percent = 0;
+    }
+
+    if (height_percent > 100) {
+        height_percent = 100;
+    }
+
+    if (height_percent < 0) {
+        height_percent = 0;
+    }
+
+    m_popupMessageData->m_xPos = g_theDisplay->Get_Width() * (width_percent / 100.0f);
+    m_popupMessageData->m_yPos = g_theDisplay->Get_Height() * (height_percent / 100.0f);
+
+    if (width < 50) {
+        width = 50;
+    }
+
+    m_popupMessageData->m_width = width;
+    m_popupMessageData->m_color = color;
+    m_popupMessageData->m_pause = pause;
+    m_popupMessageData->m_pauseMusic = pause_music;
+    m_popupMessageData->m_windowLayout = g_theWindowManager->Win_Create_Layout("InGamePopupMessage.wnd");
+    m_popupMessageData->m_windowLayout->Run_Init(nullptr);
+}
+
+void InGameUI::Message_Color(RGBColor const *color, Utf16String message, ...)
+{
+    va_list va;
+    va_start(va, message);
+    Utf16String formatted_message;
+    formatted_message.Format_VA(message, va);
+    Add_Message_Text(formatted_message, color);
+}
+
+void InGameUI::Message(Utf16String message, ...)
+{
+    va_list va;
+    va_start(va, message);
+    Utf16String formatted_message;
+    formatted_message.Format_VA(message, va);
+    Add_Message_Text(formatted_message, nullptr);
+}
+
+void InGameUI::Message(Utf8String message, ...)
+{
+    va_list va;
+    va_start(va, message);
+    Utf16String text = g_theGameText->Fetch(message.Str());
+    Utf16String formatted_message;
+    formatted_message.Format_VA(text, va);
+    Add_Message_Text(formatted_message, nullptr);
+}
+
+void InGameUI::Add_Message_Text(const Utf16String &formatted_message, const RGBColor *rgb_color)
+{
+    int color1 = m_messageColor1;
+    int color2 = m_messageColor2;
+
+    if (rgb_color != nullptr) {
+        color1 = Make_Color(0, 0, 0, 255) | rgb_color->Get_As_Int();
+        color2 = Make_Color(0, 0, 0, 255) | rgb_color->Get_As_Int();
+    }
+
+    m_uiMessages[5].full_text.Clear();

```

Comment:
use `MAX_UI_MESSAGES - 1` instead of 5. It's explicitly addressing the last element.

I suggest create

```c++
    enum
    {
        MAX_UI_MESSAGES = 6,
        LAST_UI_MESSAGE = MAX_UI_MESSAGES - 1,
    };
```

And use that.

Several times in this function.

Reply:
fixed.